src/tools/x/src/main.rs

```diff
@@ -72,26 +108,45 @@ fn main() {
         let candidate = dir.join("x.py");
 
         if candidate.exists() {
-            let mut cmd = x_command(dir);
+            let shell_script_candidate = dir.join("x");
+            if shell_script_candidate.exists() {
+                let mut cmd = x_command(dir);
 
-            cmd.args(env::args().skip(1)).current_dir(dir);
+                cmd.args(env::args().skip(1)).current_dir(dir);
 
-            let result = exec_or_status(&mut cmd);
+                let result = exec_or_status(&mut cmd);
 
-            match result {
-                Err(error) => {
-                    eprintln!("Failed to invoke `{:?}`: {}", cmd, error);
+                match result {
+                    Err(error) => {
+                        eprintln!("Failed to invoke `{:?}`: {}", cmd, error);
+                    }
+                    Ok(status) => {
+                        process::exit(status.code().unwrap_or(1));
+                    }
                 }
-                Ok(status) => {
-                    process::exit(status.code().unwrap_or(1));
+            } else {
+                // For older checkouts that do not have the x shell script, default to python
+                let mut python = Command::new(python());
+
+                python.arg(&candidate).args(env::args().skip(1)).current_dir(dir);
+
+                let result = exec_or_status(&mut python);
+
+                match result {
+                    Err(error) => {
+                        eprintln!("Failed to invoke `{}`: {}", candidate.display(), error);
+                    }
+                    Ok(status) => {
+                        process::exit(status.code().unwrap_or(1));
+                    }
                 }
             }
         }
-    }
 
-    eprintln!(
-        "x.py not found. Please run inside of a checkout of `https://github.com/rust-lang/rust`."
-    );
+        eprintln!(
+            "x.py not found. Please run inside of a checkout of `https://github.com/rust-lang/rust`."
+        );
 
-    process::exit(1);
+        process::exit(1);
+    }
```

Comment:
This changes the behaviour, it now fails immediatly, without searching the parent directories.